zkevm-circuits/src/evm_circuit/execution/return_revert.rs

```diff
@@ -333,19 +333,20 @@ impl<F: Field> ExecutionGadget<F> for ReturnRevertGadget<F> {
         }
 
         let shift = memory_offset.low_u64() % 32;
-        let memory_start_slot = memory_offset.low_u64() - shift;
-        let memory_end = memory_offset.low_u64() + length.low_u64();
-        let memory_end_slot = memory_end - memory_end % 32;
         let valid_length = if call.is_root || (call.is_create && call.is_success) {
             length.as_u64()
         } else {
-            std::cmp::min(call.return_data_length, length.as_u64())
+            call.return_data_length.min(length.as_u64())
         };
 
         let copy_rwc_inc = if valid_length == 0 {
             0
         } else {
-            (memory_end_slot - memory_start_slot) / 32 + 1
+            let begin_slot = memory_offset.low_u64() - shift;
+            let end = memory_offset.low_u64() + valid_length;
+            let end_slot = end - end % 32;
```

Comment:
This might be incorrect when `end % 32 == 0`. You could use the function [`align_range` from here](https://github.com/scroll-tech/zkevm-circuits/pull/541/files#diff-ded09b7b7016017797fa4550d232316de716e483e43ba4f5a192cceb46f36453R340). Although it is not merged yet.

Reply:
Otherwise, an alternative formula is `end = begin + length - 1`. This is the index of the last byte (not the length).

    let end = memory_offset.low_u64() + valid_length - 1;